lib/datasource/maven/index.ts

```diff
@@ -166,74 +156,83 @@ async function createUrlForDependencyPom(
-      },
-      { concurrency: 5 }
-    );
+  let workingReleaseMap: ReleaseMap = await packageCache.get<ReleaseMap>(
+    cacheNs,
+    cacheKey
+  );
 
-    artifactsInfo = results.reduce(
-      (acc, [key, value]) => ({
-        ...acc,
-        [key]: value,
-      }),
-      {}
-    );
+  if (!workingReleaseMap) {
+    workingReleaseMap = {};
 
-    // Retry earlier for status other than 404
-    const cacheTTL = Object.values(artifactsInfo).some((x) => x === null)
-      ? 60
-      : 24 * 60;
+    const unknownVersions = Object.entries(releaseMap)
+      .filter(([version, release]) => {
+        const isDiscoveredOutside = !!release;
+        const isDiscoveredInsideAndCached = !is.undefined(
+          workingReleaseMap[version]
+        );
+        const isDiscovered = isDiscoveredOutside || isDiscoveredInsideAndCached;
+        return !isDiscovered;
+      })
+      .map(([k]) => k);
+
+    if (unknownVersions.length) {
+      let retryEarlier = false;
+      const queue = unknownVersions.map(
+        (version) => async (): Promise<void> => {
+          const pomUrl = await createUrlForDependencyPom(
+            version,
+            dependency,
+            repoUrl
+          );
+          const artifactUrl = getMavenUrl(dependency, repoUrl, pomUrl);
+          const res = await isHttpResourceExists(artifactUrl);
+          const release: Release = { version };
+
+          if (is.string(res)) {
+            release.releaseTimestamp = res;
+          }
+
+          // Retry earlier for error status other than 404
+          if (res === null) {
+            retryEarlier = true;
+          }
+
+          workingReleaseMap[version] = res ? release : null;
+        }
+      );
+
+      await pAll(queue, { concurrency: 5 });
+      const cacheTTL = retryEarlier ? 60 : 24 * 60;
+      await packageCache.set(cacheNs, cacheKey, workingReleaseMap, cacheTTL);
+    }
+  }
 
-    await packageCache.set(cacheNamespace, cacheKey, artifactsInfo, cacheTTL);
+  for (const version of Object.keys(releaseMap)) {
+    releaseMap[version] ||= workingReleaseMap[version] ?? null;
   }
 
-  // Create releases for every version that exists in the repository
-  return versions
-    .filter((v) => artifactsInfo[v])
-    .map((version) => {
-      const release: Release = { version };
-      const releaseTimestamp = artifactsInfo[version];
-      if (releaseTimestamp && typeof releaseTimestamp === 'string') {
-        release.releaseTimestamp = releaseTimestamp;
-      }
-      return release;
-    });
+  return releaseMap;
 }
 
-function getFallbackReleases(versions: string[]): Release[] | null {
-  const result = versions?.map((version) => ({
-    version,
-  }));
-  return result ?? null;
+function getReleasesFromMap(releaseMap: ReleaseMap): Release[] {
+  const releases = Object.values(releaseMap).filter(Boolean);
+  if (releases.length) {
+    return releases;
+  }
+  return Object.keys(releaseMap).map((version) => ({ version }));
```

Comment:
We either return some results from HEAD responses (with further PR, to be combined with HTML-based results) or just generate from `maven-metadata.xml` without any timestamp.

The end result will be either ALL releases are having timestamp or NONE of them.